File: [6.1] Постфиксная форма/[6.1] Постфиксная форма/Stack.cpp
#include "Stack.h"

struct StackElement
{
	int value;
	StackElement *next;
};

struct Stack
{
	StackElement *head;
	int length;
};

Stack* createStack()
{
	auto stack = new Stack{ nullptr, 0 };
	return stack;
}

bool isEmpty(Stack *stack)
{
	return stack->length == 0;
}

void push(Stack *stack, int value)
{
	auto newElement = new StackElement{ value, stack->head };
	stack->head = newElement;
	++stack->length;
}

int pop(Stack *stack)
{
	if (isEmpty(stack))
	{
		return -1;
	}

	int value = stack->head->value;
	auto toDelete = stack->head;
	stack->head = stack->head->next;
	delete toDelete;
	--stack->length;
	return value;
}

void deleteStack(Stack *&stack)
{
	while (stack->head != nullptr)
	{
		auto toDelete = stack->head;
		stack->head = stack->head->next;
		delete toDelete;
	}
	delete stack;
	stack = nullptr;
}

File: semester_3/hw1/Task_1/Stack.cs
namespace Task_1
{
    /// <summary>
    /// Stack on generics
    /// </summary>
    /// <typeparam name="T">Value type</typeparam>
    public class Stack<T>
    {
        /// <summary>
        /// Stack element
        /// </summary>
        private class StackElement
        {
            /// <summary>
            /// Next stack element
            /// </summary>
            public StackElement Next { get; private set; }

            /// <summary>
            /// Element value
            /// </summary>
            public T Value { get; private set; }

            /// <summary>
            /// Constructor for stack element
            /// </summary>
            /// <param name="next">Next element</param>
            /// <param name="value">Element value</param>
            public StackElement(StackElement next, T value)
            {
                this.Next = next;
                this.Value = value;
            }
        }

        /// <summary>
        /// Stack length
        /// </summary>
        private int length;

        /// <summary>
        /// Stack head
        /// </summary>
        private StackElement head;

        /// <summary>
        /// Adding new value to stack
        /// </summary>
        /// <param name="value">Value for adding</param>
        public void Push(T value)
        {
            var newElement = new StackElement(head, value);
            head = newElement;
            ++length;
        }

        /// <summary>
        /// Removing element from stack
        /// </summary>
        /// <returns>Value from stack head</returns>
        public T Pop()
        {
            if (IsEmpty())
            {
                throw new EmptyStackException("Stack is empty!");
            }

            var temp = head.Value;
            head = head.Next;
            --length;
            return temp;
        }

        /// <summary>
        /// Getting value from stack head
        /// </summary>
        /// <returns>Value from stack head</returns>
        public T Peek()
        {
            if (IsEmpty())
            {
                throw new EmptyStackException("Stack is empty!");
            }

            return head.Value;
        }    

        /// <summary>
        /// Checking stack on emptiness
        /// </summary>
        /// <returns>True if stack is empty, otherwise false</returns>
        public bool IsEmpty() => length == 0;

        /// <summary>
        /// Getting stack length
        /// </summary>
        /// <returns>Stack length</returns>
        public int GetLength() => length;
    }
}